logtail/logtail_test.go

```diff
@@ -26,26 +30,153 @@ func TestFastShutdown(t *testing.T) {
+// accumulate some logs before the server becomes available, exercise the drain path
+func TestDrainPendingMessages(t *testing.T) {
 	ctx, cancel := context.WithCancel(context.Background())
-	uploads := 0
+	uploaded := make(chan int)
+	bodytext := ""
 	testServ := httptest.NewServer(http.HandlerFunc(
 		func(w http.ResponseWriter, r *http.Request) {
-			w.Header().Set("Content-Type", "application/json; charset=utf-8")
-			uploads += 1
+			body, err := ioutil.ReadAll(r.Body)
+			if err != nil {
+				t.Error("failed to read HTTP request")
+			}
+			bodytext += "\n" + string(body)
+
+			uploaded <- 0
 		}))
 	defer testServ.Close()
 
 	l := NewLogger(Config{BaseURL: testServ.URL}, t.Logf)
-	for i := 1; i < 10; i++ {
+	for i := 0; i < 10; i++ {
 		l.Write([]byte("log line"))
 	}
 
+	fmt.Println("server started")
+	select {
+	case <-uploaded:
+		if strings.Count(bodytext, "log line") == 10 {
+			break
+		}
+	case <-time.After(1 * time.Second):
+		t.Errorf("Timed out waiting for log uploads")
+	}
+
 	l.Shutdown(ctx)
 	cancel()
-	if uploads == 0 {
-		t.Error("no log uploads")
+	if strings.Count(bodytext, "log line") != 10 {
+		t.Errorf("want: 10 copies of \"log line\"; got: %v", bodytext)
+	}
+}
+
+func TestEncodeAndUploadMessages(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	var jsonbody []byte
+	uploaded := make(chan int)
+	testServ := httptest.NewServer(http.HandlerFunc(
+		func(w http.ResponseWriter, r *http.Request) {
+			body, err := ioutil.ReadAll(r.Body)
+			if err != nil {
+				t.Error("failed to read HTTP request")
+			}
+			jsonbody = body
+			uploaded <- 0
+		}))
+	defer testServ.Close()
+
+	l := NewLogger(Config{BaseURL: testServ.URL}, t.Logf)
+
+	// There is always an initial "logtail started" message
+	<-uploaded
+	if !strings.Contains(string(jsonbody), "started") {
+		t.Errorf("initialize: want:\"started\"; got:\"%s\"", string(jsonbody))
+	}
+
+	tests := []struct {
+		name string
+		log  string
+		want string
+	}{
+		{
+			"plain text",
+			"log line",
+			"log line",
+		},
+		{
+			"simple JSON",
+			"{\"text\": \"log line\"}",
+			"log line",
+		},
+	}
+
+	for _, tt := range tests {
+		l.Write([]byte(tt.log))
+		<-uploaded
+
+		data := make(map[string]interface{})
+		err := json.Unmarshal(jsonbody, &data)
+		if err != nil {
+			t.Error(err)
+		}
+
+		got, ok := data["text"]
+		if ok {
+			if got != tt.want {
+				t.Errorf("%s: want text \"%s\"; got \"%s\"",
+					tt.name, tt.want, got.(string))
+			}
+		} else {
+			t.Errorf("no text in: %v", data)
+		}
+
+		ltail, ok := data["logtail"]
+		if ok {
+			logtailmap := ltail.(map[string]interface{})
+			_, ok = logtailmap["client_time"]
+			if !ok {
+				t.Errorf("%s: no client_time present", tt.name)
+			}
+		} else {
+			t.Errorf("%s: no logtail map present", tt.name)
+		}
+	}
+
+	// test some special cases
+
+	// JSON log message already contains a logtail field.
+	l.Write([]byte("{\"logtail\": \"LOGTAIL\", \"text\": \"text\"}"))
```

Comment:
I'd probably use raw string literal here (backticks).

To avoid the `[]byte{` noise you can also use `io.WriteString` which can also be more efficient (not that it matters in a test):

```go
    io.WriteString(l, `{"logtail": ....}`)
```

Reply:
Done.